`test_distributions_random` failing due to change in `st.chisquare` behavior

The tests in `test_distributions_random` which rely on `pymc3_random_discrete` started failing in the `v3` branch due to changes in the behavior of `st.chisquare`. I don't think any test is making use of this helper in the `main` branch after the `v4` refactoring, so we are not seeing failures related to that here.

For instance, in the `v3` branch, `test_distributions_random.py::TestScalarParameterSamples::test_poisson` passes locally with scipy `1.6.3` but fails with `1.7.0`

It raises a `ValueError` in `st.chisquare`:

```python
            if diff_gt_tol:
                msg = (f"For each axis slice, the sum of the observed "
                       f"frequencies must agree with the sum of the "
                       f"expected frequencies to a relative tolerance "
                       f"of {rtol}, but the percent differences are:\n"
                       f"{relative_diff}")
>               raise ValueError(msg)
E               ValueError: For each axis slice, the sum of the observed frequencies must agree with the sum of the expected frequencies to a relative tolerance of 1e-08, but the percent differences are:
E               0.006036217303822937
```
This issue / breaking change was added in https://github.com/scipy/scipy/pull/13193 and mentioned in https://github.com/scipy/scipy/issues/13362 and https://github.com/scipy/scipy/issues/14298

It seems like we need to normalize the inputs as the behavior of `st.chisquare` changed.

_Originally posted by @ricardoV94 in https://github.com/pymc-devs/pymc3/pull/4808#issuecomment-869192843_

--- a/pymc3/tests/test_distributions_random.py
+++ b/pymc3/tests/test_distributions_random.py
@@ -112,7 +112,7 @@ def pymc3_random_discrete(
             if np.all(observed == expected):
                 p = 1.0
             else:
-                _, p = st.chisquare(observed+1, expected+1)
+                _, p = st.chisquare(observed + 1, expected + 1)
             f -= 1
         assert p > alpha, str(pt)
 

--- a/pymc3/tests/test_distributions_random.py
+++ b/pymc3/tests/test_distributions_random.py
@@ -107,15 +107,12 @@ def pymc3_random_discrete(
             e = ref_rand(size=size, **pt)
             o = np.atleast_1d(o).flatten()
             e = np.atleast_1d(e).flatten()
-            observed = dict(zip(*np.unique(o, return_counts=True)))
-            expected = dict(zip(*np.unique(e, return_counts=True)))
-            for e in expected.keys():
-                expected[e] = (observed.get(e, 0), expected[e])
-            k = np.array([v for v in expected.values()])
-            if np.all(k[:, 0] == k[:, 1]):
+            observed, _ = np.histogram(o, bins=min(7, len(set(o))))
+            expected, _ = np.histogram(e, bins=min(7, len(set(o))))
+            if np.all(observed == expected):
                 p = 1.0
             else:
-                _, p = st.chisquare(k[:, 0], k[:, 1])
+                _, p = st.chisquare(observed+1, expected+1)
             f -= 1
         assert p > alpha, str(pt)
 

